File: Week 2/question1.js
const check_string = (string) => {
    const newMap = new Map();

    for (let index = 0; index < string.length; index++) {
        let val = newMap.check_for_key(string[index])
        if( val === 0) {
            newMap.add_to_map(string[index])
        } else {
            newMap.increase_count(string[index]);
        }
    }

    console.log(newMap.array);
    console.log(newMap.size);

    return (
        newMap.check_lowest_count() === 1 ? 
        "No non duplicate letter" :
        `The first non duplicate character is ${newMap.check_lowest_count()}`
    )

}

class Map {
    size = -1
    array = []

    check_for_key (key) {
        if (this.size == -1)
        {
            return(0)
        }

        for (let index = 0; index <= this.size; index++) {
            if(this.array[index].key === key)
            {
                return(1)
            }            
        }

        return (0);
    }

    get_position (key) {
        for (let index = 0; index <= this.size; index++) {
            if (this.array[index].key === key){
                return (index);
            }            
        }
    }

    add_to_map(key) {
        const obj = {
            key: key,
            count: 1
        }

        this.array.push(obj);
        this.size++;
        console.log(`Key ${key} added to map!`);
    }

    increase_count (key) {
        let i = this.get_position(key);
        this.array[i].count++;
    }

    check_lowest_count () {
        for (let index = 0; index <= this.size; index++) {
            if (this.array[index].count === 1)
            {
                return (this.array[index].key)
            }          
        }

        return (1)
    }
}

console.log(check_string("minimum"))

File: Week 2/question1.c
/**
 * @file qiestion1.c
 * @author Innocent Chukwuemeka
 * @brief a function that returns the first non duplicate character in a string
 * @version 0.1
 * @date 2022-06-01
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include <stdlib.h>
#include <stdio.h>

/**
 * @brief check_string - checks string to find the first non duplicate character
 * @str: string to check
 * 
 * Return: the character
 */
typedef struct Map {
    char key;
    int count;
} mapping;

typedef struct MapDefinition {
    int size;
    mapping *array;
} map;


int get_position (mapping *arr, char key)
{
    int i = 0;
    // mapping *temp = arr.array;

    while (arr[i].key != key)
        i++;
    
    return (i);
}


int check_for_Key(map arr, char key)
{
    mapping *temp = arr.array;
    if (arr.size == 0)
        return (0);

    while(temp)
    {
        if (temp->key == key)
            return (1);
        temp++;
    }
    return (0);
}

void add_to_map(map *arr, char key)
{
    // arr->array[arr->size + 1].key = key;
    // arr->array[arr->size + 1].count = 1;
    arr->size++;
    // printf("Key %c added to map!", key);
    printf("size of arr %d", arr->size);
}

void increase_count(map *arr, char key)
{
    int position = get_position(arr->array, key);

    arr->array[position].count++;
    printf("count for Key: %c increased!", key);
}

char check_string(const char *str, int size)
{
    map newMap;

    newMap.size = -1;
    newMap.array = (mapping *)malloc(size * sizeof(mapping));

    // newMap.array[0].count = 0;
    // printf("count: %d", newMap.array[0].count);



    for (size_t i = 0; str[i] != '\0'; i++)
    {     
        if (check_for_Key(newMap, str[i]) == 0)
        {
            // add_to_map(&newMap, str[i]);
            // printf("Not in Map");
            newMap.array[newMap.size + 1].key = str[i];
            // newMap.array[newMap.size + 1].count = 1;
            printf("Key %c added to map!", str[i]);
            // newMap.size++;
            // printf("size of arr %d", arr->size);
        }
        else
        {
            // printf("in Map");
            increase_count(&newMap, str[i]);
        }
        
    }

    return ('\0');
}

int main (void)
{
    char *string = "minimum";
    char response = check_string(string, sizeof(string));

    printf("The first non duplicate character is: %c\n", response);

    return (0);
}